storage/src/lib.rs

```diff
@@ -0,0 +1,4984 @@
+        self.url_params
+            .insert("if_metageneration_not_match".into(), vec![value.into()]);
+        self
+    }
+    /// Resource name of the Cloud KMS key, of the form
+    /// projects/my-project/locations/global/keyRings/my-kr/cryptoKeys/my-key,
+    ///  that will be used to encrypt the object. Overrides the object
+    /// metadata's kms_key_name value, if any.
+    pub fn kms_key_name(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("kms_key_name".into(), vec![value.into()]);
+        self
+    }
+    /// Name of the object. Required when the object metadata is not
+    /// otherwise provided. Overrides the object metadata's name value, if
+    /// any. For information about how to URL encode object names to be path
+    /// safe, see Encoding URI Path Parts.
+    pub fn name(mut self, value: impl Into<String>) -> Self {
+        self.url_params.insert("name".into(), vec![value.into()]);
+        self
+    }
+    /// Apply a predefined set of access controls to this object.
+    pub fn predefined_acl(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("predefined_acl".into(), vec![value.into()]);
+        self
+    }
+    /// Set of properties to return. Defaults to noAcl, unless the object
+    /// resource specifies the acl property, when it defaults to full.
+    pub fn projection(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("projection".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request if the target bucket is
+    /// requester-pays bucket.
+    pub fn provisional_user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("provisional_user_project".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request. Required for Requester
+    /// Pays buckets.
+    pub fn user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("user_project".into(), vec![value.into()]);
+        self
+    }
+
+    pub async fn upload(
+        mut self,
+        media: impl Into<BytesReader>,
+        media_mime_type: impl Into<std::string::String>,
+    ) -> Result<model::Object> {
+        let client = self.client.inner;
+        let tok = client.cred.access_token().await.map_err(Error::wrap)?;
+        let body = serde_json::to_vec(&self.request).map_err(Error::wrap)?;
+        let form = reqwest::multipart::Form::new()
+            .part(
+                "body",
+                reqwest::multipart::Part::bytes(body)
+                    .mime_str("application/json")
+                    .map_err(Error::wrap)?,
+            )
+            .part(
+                "media",
+                reqwest::multipart::Part::bytes(media.into().read_all().await?.as_ref().to_owned())
```

Comment:
I'm surprised to see `.read_all()` here.  What if the caller wants to upload a 300GB file?

Reply:
This is the exact reason we eventually need to support chucked uploads. Then smaller buffers can be uploaded over several requests. The way this method is structured we could bolt on an optional call param called "chunk_size". If set it would do a buffered chuncked upload instead of the single shot upload as shown here. I still need to do more investigation into this space as well.

Reply:
👍